ROX-17191: roxctl connectivity-map supports route resources

shireenf-ibm opened the pull request:
## Description
- update to netpol-analyzer v0.3.2 :
  - removed connections between workload to itself from the report
  - support Ingress/Route analysis - include connections inferred from such resources

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
Unit test and bats tests where added

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

shireenf-ibm committed:
initial_support

shireenf-ibm committed:
adding bats tests for netpol-analyzer v0.3.1

shireenf-ibm committed:
adding acs-security-demo test

shireenf-ibm committed:
eof fixes

shireenf-ibm committed:
fixing large file lint err; removed unnecessary info

shireenf-ibm committed:
reducing tests

shireenf-ibm committed:
update version

adisos committed:
update documentation

Signed-off-by: adisos <adisos@il.ibm.com>

adisos committed:
additional documentation update

Signed-off-by: adisos <adisos@il.ibm.com>

janisz reviewed (commented):
Could you split this PR into bump and the rest

shireenf-ibm committed:
adding expected_output file to the new unit test + fix existing outpust to match new version

vikin91 reviewed (commented):
Thank you for the PR! It may take me a moment until I give it a proper look, but I am sharing some initial thoughts already.

1. Would be good to rebase to incorporate recent test improvement #7154 and changes that are required to pass CI checks
2. Is there any way to avoid duplication of the `testdata/acs-secuirty-demos`? Could the bats-tests access the test-data of unit tests maybe?

shireenf-ibm committed:
Update roxctl/connectivity-map/README.md

Co-authored-by: Piotr Rygielski <114479+vikin91@users.noreply.github.com>

shireenf-ibm committed:
removed duplicated tests dirs , and updated bats tests dir paths to unit-tests-dir-paths

shireenf-ibm commented:
> Thank you for the PR! It may take me a moment until I give it a proper look, but I am sharing some initial thoughts already.
> 
>     1. Would be good to rebase to incorporate recent test improvement [test(netpol): check netpol generated output #7154](https://github.com/stackrox/stackrox/pull/7154) and changes that are required to pass CI checks
> 
>     2. Is there any way to avoid duplication of the `testdata/acs-secuirty-demos`? Could the bats-tests access the test-data of unit tests maybe?

1. rebased
2. done, removed tests directories under `tests/roxctl/bats-tests/test-data/np-guard/` which are similar to other directories under `connectivity-map\testdata `, and updated the paths in the bats tests files

janisz commented on go.mod:
```diff
@@ -70,7 +70,7 @@ require (
 	github.com/moby/sys/mountinfo v0.6.2
 	github.com/np-guard/cluster-topology-analyzer v1.7.0
-	github.com/np-guard/netpol-analyzer v0.2.2
+	github.com/np-guard/netpol-analyzer v0.3.2
```
- https://github.com/stackrox/stackrox/pull/7316
 
 ```suggestion
	github.com/np-guard/netpol-analyzer v0.4.0
```

vikin91 replied:
I see that Dependabot thinks that `0.y+1` is compatible with `0.y`. This would be true if that is `1.y`, but there is no compatibility guarantee between the `0.y` versions. https://semver.org/#spec-item-4

Can we somehow configure it to respect that?

adisos replied:
This PR  should remain with `v0.3.2` of `netpol-analyzer`.
We have a new release of `v0.4.2` for which we will open soon a separate PR (on top of this PR/branch), since it adds support for the `connectivity-diff` command.

vikin91 commented:
Another one for the future:
Usage of `--focus-workload` needs to be better documented in the docs and in the code. Here few examples of what I tried:

1. Non-existing focus workload: 
```
bin/darwin_amd64/roxctl connectivity-map roxctl/connectivity-map/testdata/acs-security-demos --focus-workload=abc
WARN:	This is a Technology Preview feature. Red Hat does not recommend using Technology Preview features in production.
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/checkout/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/recommendation/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/reports/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/asset-cache/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/webapp/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 1, skipping object with type: RoleBinding
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 1, skipping object with type: ClusterRoleBinding
{ingress-controller} => frontend/asset-cache[Deployment] : TCP 8080
{ingress-controller} => frontend/webapp[Deployment] : TCP 8080
```

I expected either an empty output or a warning that this workload does not exist.

2. Workload with namespace

```
➜ bin/darwin_amd64/roxctl connectivity-map roxctl/connectivity-map/testdata/acs-security-demos --focus-workload=payments/gateway
WARN:	This is a Technology Preview feature. Red Hat does not recommend using Technology Preview features in production.
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/checkout/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/recommendation/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/reports/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/asset-cache/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/webapp/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 1, skipping object with type: RoleBinding
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 1, skipping object with type: ClusterRoleBinding
{ingress-controller} => frontend/asset-cache[Deployment] : TCP 8080
{ingress-controller} => frontend/webapp[Deployment] : TCP 8080
```

Again, no hint that the user did something wrong. The two lines with `{ingress-controller}` may suggest that this has something to do with the workload I provided.

3. Correct usage of this flag

```
➜ bin/darwin_amd64/roxctl connectivity-map roxctl/connectivity-map/testdata/acs-security-demos --focus-workload=gateway
WARN:	This is a Technology Preview feature. Red Hat does not recommend using Technology Preview features in production.
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/checkout/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/recommendation/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/backend/reports/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/asset-cache/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/frontend/webapp/configmap.yaml, document: 0, skipping object with type: ConfigMap
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/mastercard-processor/deployment.yaml, document: 1, skipping object with type: RoleBinding
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 0, skipping object with type: ServiceAccount
INFO:	in file: roxctl/connectivity-map/testdata/acs-security-demos/payments/visa-processor/deployment.yaml, document: 1, skipping object with type: ClusterRoleBinding
backend/checkout[Deployment] => payments/gateway[Deployment] : TCP 8080
payments/gateway[Deployment] => payments/mastercard-processor[Deployment] : TCP 8080
payments/gateway[Deployment] => payments/visa-processor[Deployment] : TCP 8080
{ingress-controller} => frontend/asset-cache[Deployment] : TCP 8080
{ingress-controller} => frontend/webapp[Deployment] : TCP 8080
```
Now the workload has been found, but why we see the `{ingress-controller}` again?

I would suggest to open a bug ticket in JIRA and address that in a separate PR.

vikin91 reviewed (approved):
Looks good provided that https://github.com/stackrox/stackrox/pull/7173#issuecomment-1671704756 will be addressed in a follow-up before the release.

adisos commented:
> Looks good provided that [#7173 (comment)](https://github.com/stackrox/stackrox/pull/7173#issuecomment-1671704756) will be addressed in a follow-up before the release.

Will address those issues in a separate PR.

vikin91 commented:
Thanks! I created a Jira for this: https://issues.redhat.com/browse/ROX-19015
